File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    //lec 9 method 4 using tabulation
    int coinChange(vector<int>& coins, int amount) {
           int *dp=new int[amount+1];
        for(int i=0;i<amount+1;i++){
            dp[i]=INT_MAX;//-1 bhi rakh sakte the
        }
        dp[0]=0;
        for(int i=1;i<=amount;i++){
            for(int coin:coins){
                if(i-coin>=0 && dp[i-coin]!=INT_MAX){
                    dp[i]=min(dp[i-coin]+1,dp[i]);
                }
            }
        }
        return dp[amount]==INT_MAX?-1:dp[amount];
    }
};

File: 322-coin-change/322-coin-change.java
//lec9
//earlier method was easier
// Memoization - Approach 2
// Time - O(Amount * Coins), Space - O(Amount * Coins) for DP table
// Recursion Call Stack -> O(Coins) Height, O(Amount) Calls/Breadth
//we are considering whether coin will come or not and if it will come should we take 1 or 2 coins or 3 coins etc
//2 variable states therefore 2 base cases
class Solution {
    
        public int memo(int amount, int idx, int[] coins, int[][] dp) {
        if (amount == 0)
            return 0;
        if (idx == coins.length)
            return Integer.MAX_VALUE;
        if (dp[amount][idx] != -1)
            return dp[amount][idx];

        int minCoins = Integer.MAX_VALUE;
        for (int coin = 0; amount >= coins[idx] * coin; coin++) {
            int ans = memo(amount - coins[idx] * coin, idx + 1, coins, dp);
            if (ans < Integer.MAX_VALUE)
                ans += coin;
            minCoins = Math.min(minCoins, ans);
        }

        return dp[amount][idx] = minCoins;
    }
    
    public int coinChange(int[] coins, int amount) {
                int[][] dp = new int[amount + 1][coins.length];
        for (int i = 0; i <= amount; i++) {
            for (int j = 0; j < coins.length; j++) {
                dp[i][j] = -1;
            }
        }

        int ans = memo(amount, 0, coins, dp);
        if (ans == Integer.MAX_VALUE)
            return -1;
        return ans;
    }
}